src/openms/source/ANALYSIS/OPENSWATH/IonMobilityScoring.cpp

```diff
@@ -474,4 +648,189 @@ namespace OpenMS
+      identification_mobilogram.setName(transition[0].getNativeID());
+      mobilograms.push_back(identification_mobilogram);
+
+      // Check to make sure IM of identification is not -1, otherwise assign 0 for scores
+      if ( identification_im!=-1 )
+      {
+        std::vector<double> im_grid = computeGrid_(mobilograms, eps); // ensure grid is based on all profiles!
+        mobilograms.pop_back();
+
+        // Step 3: Align the IonMobilogram vectors to the grid
+        std::vector< OpenMS::Mobilogram > aligned_mobilograms;
+        for (const auto &mobilogram : mobilograms)
+        {
+          Mobilogram aligned_mobilogram;
+          Size max_peak_idx = 0;
+          alignToGrid_(mobilogram, im_grid, aligned_mobilogram, eps, max_peak_idx);
+          aligned_mobilogram.setName(mobilogram.getName());
+          aligned_mobilograms.push_back(std::move(aligned_mobilogram));
+        }
+
+        size_t left = 0, max = 0, right = 0;
+        if ( apply_im_peak_picking ) 
+        {
+          if ( !aligned_mobilograms.empty())
+          {
+            OpenMS::Mobilogram summed_mobilogram = sumAlignedMobilograms(aligned_mobilograms);
+
+            PeakPickerMobilogram picker_;
+            Param picker_params = picker_.getParameters();
+            picker_params.setValue("method", "corrected");
+            picker_.setParameters(picker_params);
+            Mobilogram picked_mobilogram, smoothed_mobilogram;
+            picker_.pickMobilogram(summed_mobilogram, picked_mobilogram, smoothed_mobilogram);
+            std::tie(max, left, right) = findHighestPeak(picker_);
+            scores.im_drift_left = im_grid[left];
+            scores.im_drift_right = im_grid[right];
+            filterPeakIntensities(aligned_mobilograms, left, right);
+
+            OPENMS_LOG_DEBUG << "  -- IM peak picking for summed mobilograms for " << summed_mobilogram.getName() << " found peak at " << im_grid[max] << "(" << im_grid[left] << " - " << im_grid[right] << ")" << std::endl;
+          }
+          else
+          {
+            scores.im_drift_left = -1;
+            scores.im_drift_right = -1;
+          }
+        }
+
+        Mobilogram aligned_identification_mobilogram;
+        Size max_peak_idx = 0;
+        alignToGrid_(identification_mobilogram,
+                    im_grid,
+                    aligned_identification_mobilogram,
+                    eps,
+                    max_peak_idx);
+        aligned_identification_mobilogram.setName(transition[0].getNativeID());
+        if ( apply_im_peak_picking )
+        {
+          if ( !aligned_identification_mobilogram.empty())
+          {
+            filterPeakIntensities(aligned_identification_mobilogram, left, right);
+          }
+        }
+
+        std::vector< std::vector< double > > aligned_int_vec;
+        extractIntensities(aligned_mobilograms, aligned_int_vec);
```

Comment:
an eigen matrix would have a better memory alignment than a vector of vectors as it internally can store everything in one contiguous memory block. a vector of vector will allocate several regions (one for each vector) on the heap (simply speaking)

Reply:
I think we would have to adjust/refactor the Xcorr methods as well then, because they take in a vector of vectors, which some of them do need to be re-visited anyways. I think this could be in a separate PR though?